utils/waitable_counter.hh

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2021-present ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <map>
+#include <iosfwd>
+#include <exception>
+
+#include <seastar/core/future.hh>
+#include <seastar/core/future-util.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/condition-variable.hh>
+
+namespace utils {
+
+/**
+ * Simple counter type wrapper, that allows
+ * waiting for it to reach a given value.
+ * 
+ * Actual content must be monotonically 
+ * increasing. 
+ * 
+ * Waiting using future<> wait routines
+ * is less efficient than using coroutine
+ * when(..) calls, as the former can cause
+ * spurious wakeups wheras the latter will 
+ * not.
+ */
+template<typename T, typename Cmp = std::less<T>>
+class waitable_counter {
+private:
+    T _value;
+    condition_variable _cond;
+    Cmp _cmp;
+
+    template<std::invocable<> Func>
+    waitable_counter& set(T v, Func&& f) {
+        if (_cmp(v, _value)) {
+            throw std::invalid_argument("Values must be monotonically increasing");
+        }
+        _value = std::move(v);
+        f();
```

Comment:
What is this f() call for? it seams pretty specific.
If it has some general part maybe it is worthwhile to document the API.